File: sdk-python/tests/iwfcompat/workflow_uncompleted_test.py
# Copyright (c) 2026 Super Durable, Inc.
#
# Licensed under the Super Durable Source License 1.0.
# You may not use this file except in compliance with the License.
# See the LICENSE file in the repository root.
#
# SPDX-License-Identifier: LicenseRef-Super-Durable-1.0

from datetime import timedelta

from dex import Client, StartFlowOptions, StopFlowOptions, StopType

from . import iwf_flows


def compile_wait_and_flow_timeouts(client: Client) -> None:
    options = StartFlowOptions(timeout=timedelta(seconds=1))
    client.start_flow(iwf_flows.SIGNAL, "wait-timeout", 0, options)
    output: int = client.wait_for_flow(
        "wait-timeout",
        int,
        timedelta(milliseconds=1),
    )
    del output


def compile_cancellation_termination_and_failure(client: Client) -> None:
    client.stop_flow("cancel")
    client.stop_flow(
        "terminate",
        StopFlowOptions(StopType.TERMINATE, "terminated"),
    )
    client.stop_flow(
        "fail",
        StopFlowOptions(StopType.FAIL, "failed by API"),
    )


def compile_worker_failure_modes(client: Client) -> None:
    client.start_flow(iwf_flows.FORCE_FAIL, "force-fail", 0)
    client.start_flow(iwf_flows.STATE_FAILURE, "state-failure", 0)
    client.start_flow(iwf_flows.STATE_TIMEOUT, "state-timeout", 0)
    client.start_flow(iwf_flows.EMPTY_DECISION, "empty-decision", 0)

File: sdk-typescript/test/integ/workflow_uncompleted_test.ts
// Copyright (c) 2026 Super Durable, Inc.
//
// Licensed under the Super Durable Source License 1.0.
// You may not use this file except in compliance with the License.
// See the LICENSE file in the repository root.
//
// SPDX-License-Identifier: LicenseRef-Super-Durable-1.0

import { StopType, doubleCodec, type Client } from "../../src/index.js";

import * as flows from "./iwf_flows.js";

export async function compileWaitAndFlowTimeouts(client: Client): Promise<void> {
  await client.startFlow(flows.SIGNAL, "wait-timeout", 0, { timeoutMs: 1_000 });
  const output: number = await client.waitForFlow(
    "wait-timeout",
    doubleCodec,
    1,
  );
  void output;
}

export async function compileCancellationTerminationAndFailure(
  client: Client,
): Promise<void> {
  await client.stopFlow("cancel");
  await client.stopFlow("terminate", {
    type: StopType.TERMINATE,
    reason: "terminated",
  });
  await client.stopFlow("fail", {
    type: StopType.FAIL,
    reason: "failed by API",
  });
}

export async function compileWorkerFailureModes(client: Client): Promise<void> {
  await client.startFlow(flows.FORCE_FAIL, "force-fail", 0);
  await client.startFlow(flows.STATE_FAILURE, "state-failure", 0);
  await client.startFlow(flows.STATE_TIMEOUT, "state-timeout", 0);
  await client.startFlow(flows.EMPTY_DECISION, "empty-decision", 0);
}
